icu4c/source/common/rbbitblb.cpp

```diff
@@ -1549,20 +1578,38 @@ void RBBITableBuilder::exportSafeTable(void *where) {
         return;
     }
 
-    table->fRowLen    = offsetof(RBBIStateTableRow, fNextState) + sizeof(uint16_t) * catCount;
     table->fNumStates = fSafeTable->size();
+    bool use8Bits = table->fNumStates < kStateThresholdFor8BitsTable;
+    if (use8Bits) {
+        table->fRowLen    = offsetof(RBBIStateTableRowS8, fNextState) + sizeof(uint8_t) * catCount;
+    } else {
+        table->fRowLen    = offsetof(RBBIStateTableRowS16, fNextState) + sizeof(int16_t) * catCount;
+    }
     table->fFlags     = 0;
+    if (use8Bits) {
+        table->fFlags  |= RBBI_8BITS_ROWS;
+    }
     table->fReserved  = 0;
 
     for (state=0; state<table->fNumStates; state++) {
         UnicodeString *rowString = (UnicodeString *)fSafeTable->elementAt(state);
         RBBIStateTableRow   *row = (RBBIStateTableRow *)(table->fTableData + state*table->fRowLen);
-        row->fAccepting = 0;
-        row->fLookAhead = 0;
-        row->fTagIdx    = 0;
-        row->fReserved  = 0;
-        for (col=0; col<catCount; col++) {
-            row->fNextState[col] = rowString->charAt(col);
+        if (use8Bits) {
+            row->s8.fAccepting = 0;
+            row->s8.fLookAhead = 0;
+            row->s8.fTagIdx    = 0;
+            row->s8.fReserved  = 0;
+            for (col=0; col<catCount; col++) {
+                row->s8.fNextState[col] = (uint8_t)rowString->charAt(col);
```

Comment:
This might be a good place for a U_ASSERT that the value being stored is in range. Should be < the number of states.